src/models/room.js

```diff
@@ -207,7 +207,12 @@ Room.prototype.getLiveTimeline = function() {
  */
 Room.prototype.resetLiveTimeline = function(backPaginationToken) {
     for (let i = 0; i < this._timelineSets.length; i++) {
-        this._timelineSets[i].resetLiveTimeline(backPaginationToken);
+        // Flush out non-live timelines. We do this to reduce the amount of memory
+        // used, as storing multiple distinct copies of room state (each of which
+        // can be 10s of MBs) for each timeline is expensive. This is particularly
+        // noticeable when the client unsleeps and there are a lot of 'limited: true'
+        // responses. https://github.com/vector-im/riot-web/issues/3307#issuecomment-282895568
+        this._timelineSets[i].resetLiveTimeline(backPaginationToken, true);
```

Comment:
> a change that deliberately leaves the datastructures in an inconsistent state is a bad idea.

I'm not, I'm removing all the events in the timeline then adding the ones I got from a `limited` sync. Where is the inconsistency? The event ID to timeline map is cleared when we `flush` to preserve consistency: https://github.com/matrix-org/matrix-js-sdk/blob/master/src/models/event-timeline-set.js#L163

> it shouldn't jump down if you're viewing historical events.

Why? This strictly happens on a `limited` /sync only, which restricts this to startup/load/unsleep. If we *don't* jump down, then the events needs to hang around until some indeterminate point in time when they can be reaped, and it's unclear when that should happen or how that should work, and sounds exceptionally prone to leaking the events beyond the point where you've removed them from the JS SDK object model. Your examples focus on the assumption that the events are still around, and the entire point of this PR is to make them NOT hang around.

If they don't hang around, then it's really quite simple: we drop all events from the room and only populate the most recent 20 messages. The things to check are:
 - Make sure that the other events are GCed.
 - Make sure that if you were viewing some of the older events, that the screen jumps to the bottom as they no longer exist.

Both of those cases I have manually tested and confirmed. I think we may have philosophical differences on how this should behave, so may need @ara4n's opinion on how to proceed.